pwndbg/dbg/lldb/__init__.py

```diff
@@ -586,6 +588,66 @@ def set_enabled(self, enabled: bool) -> None:
         self.inner.SetEnabled(enabled)
 
 
+class OneShotAwaitable:
+    """
+    Used as part of the logic for the execution controller. This is an Awaitable
+    object that yields the value passed to its constructor exactly once.
+    """
+
+    def __init__(self, value: Any):
+        self.value = value
+
+    def __await__(self) -> Generator[Any, Any, None]:
+        yield self.value
+
+
+class YieldContinue:
+    """
+    Continues execution of the process until the breakpoint or watchpoint given
+    in the constructor is hit or the operation is cancelled.
+
+    This class is part of the execution controller system, so it is intented to
+    be yielded by the async function with access to an execution controller, and
+    caught and hanlded by the event loop in the LLDB Pwndbg CLI.
+    """
+
+    target: LLDBStopPoint
+
+    def __init__(self, target: LLDBStopPoint):
+        self.target = target
+
+
+class YieldSingleStep:
+    """
+    Moves execution of the process being debugged forward by one instruction.
+
+    This class is part of the execution controller system, so it is intented to
+    be yielded by the async function with access to an execution controller, and
+    caught and hanlded by the event loop in the LLDB Pwndbg CLI.
+    """
+
+    pass
+
+
+class LLDBExecutionController(pwndbg.dbg_mod.ExecutionController):
+    @override
+    def single_step(self) -> Awaitable[None]:
+        return OneShotAwaitable(YieldSingleStep())
+
+    @override
+    def cont(self, target: pwndbg.dbg_mod.StopPoint) -> Awaitable[None]:
+        assert isinstance(target, LLDBStopPoint)
+        t: LLDBStopPoint = target
```

Comment:
Just pass `target`? Can't we handle this with typing better?

Reply:
I was gonna say "it's probably either that or `typing.cast()`, which is even worse", but apparently MyPy is smart enough now that we don't need either of them, and just having the assertion is good enough for it to catch on. I'm gonna take the assignment out.